docs/DeveloperGuide.md

```diff
@@ -2,14 +2,372 @@
+   for command arguments, allowing subclasses to generate specific **`Command` instances** based on **command keywords
+   and
+   parsed arguments.**
+   <br />
+   <br />
+   The `Parser` class maintains a list of these `CommandMetadata` subclasses instances and iterates through them to
+   identify a given user command.
+   <br />
+   <br />
+   Below is a diagram that shows this relationship
+
+   ![ParserClassDiagram.png](diagrams/ParserClassDiagram.png)
+   <br />
+   <br />
+   **Use of regular expressions (Regex) in FAP:**
+
+   The add command class requires a user input that best
+   matches this string
+    ```
+    add c/COURSECODE w/SEMESTERTAKEN
+    ```
+   where `COURSECODE` and `SEMESTERTAKEN` have their defined restrictions: `COURSECODE` should best match an actual
+   course code at NUS, `SEMESTERTAKEN` should be a number value in some range. The `COURSECODE` and `SEMESTERTAKEN`
+   will thus have their own argument regex pattern.
+   <br />
+   <br />
+   A simple example would be that `SEMESTERTAKEN` would be a number ranging from 1-8 to represent a normal honours
+   pathway for a CEG student (FAP's target user). A regex pattern for that would look like `w/(?<semester>[1-8])`.
+   An **argument name capturing group** `semester` is enclosed within the brackets so that the argument group will be
+   **named** and thus the argument value (anywhere between `1-8`) can be referenced/called by using the
+   `matcher.group()` method.
+   Meanwhile, the `Pattern` and `Matcher` methods used for regex would handle the checks that the argument value given
+   is indeed between `1-8`
+   <br />
+   <br />
+   A userInput regex would thus follow this convention:
+    ```
+    keyword argument_1 argument_2 ...
+    ```
+   This full regex pattern for a command itself can be generated by having a `keyword`, as well as all the
+   `argument group names` (a name to use so as to _reference_ the argument) and the `argument regex pattern`
+   corresponding to that name reference. Typically, these arguments would be spaced out and thus a `\s+` (representing
+   at least one whitespace character) is placed between the gaps of the regex pattern for
+   `keyword, argument_1, argument_2...`
+   <br />
+   <br />
+   While regex allows the `userInput` checks to be prudent, as well as potentially offering the flexibility for string
+   inputs to allow a different order of arguments, there are limitations where it becomes hard to determine the exact
+   error of the user's input solely based on the regular expressions, because it solely returns a true/false value
+   if the string value itself fits the criteria given). Regardless, we think the use of regex in FAP can help provide
+   us **safety in the arguments** that passes through to the commands via the userInput.
+   <br />
+   <br />
+   **Developer usage FAP: Parser & CommandMetadata class as of v2.0**: **How to create a new command**
```

Comment:
Great job in going the extra mile in describing the way to maintaining and adding to the codebase for future developers!